File: top_100_liked/longest_palindromic_substring.py
class Solution:
    ### Expand Center
    def AlongestPalindrome(self, s: str) -> str:
        maxx, start, end = 0, 0, 0

        for i in range(len(s)-1):
            # even palindrome center
            maxx = max(maxx, self.expandCenter(s, i, i+1))
            if(end - start + 1 < maxx):
                start = i - maxx//2 + 1
                end   = i + maxx//2
            if(i - 1 >= 0 and i + 1 < len(s)):
                # odd palindrome center
                maxx = max(maxx, self.expandCenter(s, i - 1, i + 1) + 1)
                if(end - start + 1 < maxx):
                    start = i - maxx//2
                    end   = i + maxx//2

        return s[start:end+1]

    def expandCenter(self, s, i, j):
        count = 0
        while(i >= 0 and j < len(s)):
            if(s[i] == s[j]):
                count += 2
                i -= 1
                j += 1
            else:
                break
        return count

    # DP
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        # edge case
        if(n <= 1):
            return s
        # nxn matrix
        DP = [[0]*n for _ in range(n)]
        start, end = 0, 0
        # initial case
        for i in range(n):
            DP[i][i] = 1
        for i in range(n-1):
            if(s[i] == s[i+1]):
                DP[i][i+1] = 1
                if(end - start < 1):
                    start, end = i, i+1
        # constructing table
        for j in range(1, n):
            for i in range(0, j):
                if(DP[i+1][j-1] and s[i] == s[j]):
                    DP[i][j] = 1
                    if(end - start < j - i):
                        start, end = i, j
        return s[start:end+1]

File: top_100_liked/cpp/longest_palindromic_substring.cpp
class Solution {
public:
    struct sub_index{
        size_t i, j;    
    };
    
    string longestPalindrome(string s) {
        //return expandFromCenter(s); 
        return dynamicProgramming(s);
    }
    
    string expandFromCenter(string s){
        size_t maxx = 0;
        sub_index ans;
        // odd center
        size_t s_len = s.length();
        for(size_t i = 0; i < s_len; ++i){
            for(size_t j = 0; j < s_len; ++j){
                if(i < j || (i + j) >= s_len)
                    break;
                if(s[i-j] != s[i+j])
                    break;
                else if(2*j + 1 > maxx){
                    maxx = 2 * j + 1;
                    ans.i = i - j;
                    ans.j = i + j;
                }
            }
        }
        // even center
        for(size_t i = 0; i < s_len; ++i){
            for(size_t j = 0; j < s_len; ++j){
                if(i < j || (i + j + 1) >= s_len)
                    break;
                if(s[i-j] != s[i+j+1])
                    break;
                else if(2*j + 2 > maxx){
                    maxx = 2*j + 2;
                    ans.i = i - j;
                    ans.j = i + j + 1;
                }
            }
        }
                
        return s.substr(ans.i, ans.j - ans.i + 1);
    }
    
    string dynamicProgramming(string s){
        size_t n = s.length(), maxx = 0;
        sub_index ans;
        int dp[n][n];
        for(int i = 0; i < n; ++i)
            for(int j = 0; j < n; ++j)
                dp[i][j] = 0;
        
        ans.i = 0;
        ans.j = 0;
        // init dp table
